Add Drag Gesture for StageMode for Model View

Phipson opened the pull request:
#### b83e18a1c34ddb557a9730f71fd6155bb05fd89d
<pre>
Add Drag Gesture for StageMode for Model View
<a href="https://bugs.webkit.org/show_bug.cgi?id=287048">https://bugs.webkit.org/show_bug.cgi?id=287048</a>
<a href="https://rdar.apple.com/137779076">rdar://137779076</a>

Reviewed by Wenson Hsieh and Ada Chan.

This PR is the second part of 3 code changes used to introduce stagemode interactions to the model view.
Specifically, this code change adds a UIPanGestureRecognizer that will be used to update the entityTransform
for the model view in <a href="https://rdar.apple.com/141251753">rdar://141251753</a>. When the stagemode attribute is set to orbit, and we have hit an
HTMLModelElement, the gesture will feed transform updates to the model player, which will use those transforms
to update the transform of the model. We wrap the model inside an interaction container, to separate gesture-
based transform updates from JavaScript/Animation-driven transform updates. Also, transform supplied by the
gesture will be fed into a StageModeInteractionDriver that will alter the transform of the model, depending
on the stagemode attribute that has been set for the element.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::supportsStageModeInteraction const):
(WebCore::HTMLModelElement::updateStageModeTransform):
(WebCore::HTMLModelElement::endStageModeInteraction):
- Add API calls to determine if stagemode is defined for the model; to update the model via the model player;
and end the interaction via the model player once the gesture terminates

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::supportsStageModeInteraction const):
(WebCore::HTMLModelElement::beginStageModeTransform):
(WebCore::HTMLModelElement::updateStageModeTransform):
(WebCore::HTMLModelElement::endStageModeInteraction):
- Add API calls to determine if stagemode is defined for the model; to update the model via the model player;
and end the interaction via the model player once the gesture terminates

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::beginStageModeTransform):
(WebCore::ModelPlayer::updateStageModeTransform):
(WebCore::ModelPlayer::endStageModeInteraction):
- Defines the virtual and default behavior of the model player to handle stage mode updates from the gesture;
to be implemented by the ModelProcessModelPlayer and ModelProcessModelPlayerProxy

* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::requestInteractiveModelElementAtPoint):
(WebCore::EventHandler::stageModeSessionDidUpdate):
(WebCore::EventHandler::stageModeSessionDidEnd):
- Handles hit-testing and querying the HTMLModelElement (if available) to perform the stagemode interaction

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::beginStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::endStageModeInteraction):
- Implements the transform updates directly on the model in the model process, to be done in <a href="https://bugs.webkit.org/show_bug.cgi?id=287820">https://bugs.webkit.org/show_bug.cgi?id=287820</a>

* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:
* Source/WebKit/UIProcess/Model/StageModeInteractionState.h: Added.
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestInteractiveModelElementAtPoint):
(WebKit::WebPageProxy::stageModeSessionDidUpdate):
(WebKit::WebPageProxy::stageModeSessionDidEnd):
- Passes calls from the WKContentViewInteraction gesture to the WebPage in the Web Process

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didReceiveInteractiveModelElement):
- Passes the result of the initial query when the gesture begins back to the content view to update its interaction state

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _removeDefaultGestureRecognizers]):
(-[WKContentView _addDefaultGestureRecognizers]):
(-[WKContentView gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:]):
(-[WKContentView gestureRecognizerShouldBegin:]):
(-[WKContentView _modelInteractionPanGestureRecognized:]):
(-[WKContentView modelInteractionPanGestureDidBeginAtPoint:]):
(-[WKContentView modelInteractionPanGestureDidUpdateWithPoint:]):
(-[WKContentView modelInteractionPanGestureDidEnd]):
(-[WKContentView didReceiveInteractiveModelElement:]):
(-[WKContentView cleanUpStageModeSessionState]):
(-[WKContentView updateStageModeSessionTransformWithPoint:]):
- Defines functions to add the model-specific UIPanGestureRecognizer and maintain an internal interaction state of the stagemode
interaction

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::didReceiveInteractiveModelElement):
- Receives the hit-testing result from the Web Process and routes calls to the PageClient

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::beginStageModeTransform):
(WebKit::ModelProcessModelPlayer::updateStageModeTransform):
(WebKit::ModelProcessModelPlayer::endStageModeInteraction):
- Routes calls from the HTMLModelElement to the ModelProcessModelPlayerProxy

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestInteractiveModelElementAtPoint):
(WebKit::WebPage::stageModeSessionDidUpdate):
(WebKit::WebPage::stageModeSessionDidEnd):
- Receives the stagemode message from the UI Process and hands it off to the EventHandler

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::positionInformation):
- Adds state information about whether there is a model element with stagemode enabled at the position we are hit-testing

Canonical link: <a href="https://commits.webkit.org/290687@main">https://commits.webkit.org/290687@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6585b01ab1e31d8be1c8c882cb45077a3add682f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90811 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10348 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95830 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/41604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92864 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10742 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18665 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/41604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93812 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/50174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36710 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40726 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37778 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/97684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18006 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/97684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18265 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78162 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/97684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19281 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22515 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18015 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/23360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17754 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21210 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19538 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -3848,6 +3849,11 @@ static void animationPositionInformation(WebPage& page, const InteractionInforma
 
+#if ENABLE(MODEL_PROCESS)
+    if (RefPtr modelElement = dynamicDowncast<HTMLModelElement>(hitTestNode))
+        info.isInteractiveModel = modelElement->model() != nullptr && modelElement->supportsStageModeInteraction();
```
WebKit style prefers implicit nullity checks.

```suggestion
        info.isInteractiveModel = modelElement->model() && modelElement->supportsStageModeInteraction();
```

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1274,6 +1275,12 @@ class WebPage final : public API::ObjectImpl<API::Object::Type::BundlePage>, pub
 
+#if ENABLE(MODEL_PROCESS)
+    void requestInteractiveModelElementAtPoint(WebCore::IntPoint clientPosition);
+    void stageModeSessionDidUpdate(std::optional<WebCore::ElementIdentifier>, WebCore::TransformationMatrix);
```
```suggestion
    void stageModeSessionDidUpdate(std::optional<WebCore::ElementIdentifier>, const WebCore::TransformationMatrix&);
```

Nit - probably `const&` or `&&` here to prevent unnecessary copies of a (somewhat-large) set of values.

pxlcoder commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -7702,13 +7702,15 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = BC90957F12554CF900083756 /* DebugRelease.xcconfig */;
 			buildSettings = {
+				COMPILER_INDEX_STORE_ENABLE = NO;
```
Undo.

pxlcoder commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -7702,13 +7702,15 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = BC90957F12554CF900083756 /* DebugRelease.xcconfig */;
 			buildSettings = {
+				COMPILER_INDEX_STORE_ENABLE = NO;
```
Undo.

rr-codes commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,110 @@
+// Copyright (C) 2024 Apple Inc. All rights reserved.
```
why does this need to be in WebKitSwift?

Phipson replied:
So the original intention was that we would adopt a specific interaction for visionOS using an API that is written exclusively in Swift, meaning that we needed a wrapper to use it with the model player. 

Although we have deferred that adoption for now, we have recently received design guidance to adopt some animation behaviors as part of the stagemode interaction which requires using Swift-only RealityKit APIs like

https://developer.apple.com/documentation/realitykit/hastransform/move(to:relativeto:duration:timingfunction:)-35qp2

and the Observable API https://developer.apple.com/documentation/observation/withobservationtracking(_:onchange:)

both of which are Swift-exclusive. We have a follow-up task/bug to adopt that, but this will effectively set us up for that effort.

So as a means to both adopt the said interaction in future, and to support the animations in a subsequent code change, we've opted to implement this in Swift (and therefore WebKitSwift)

rr-codes replied:
I meant moreso, this can be in WebKit itself, and not WebKitSwift

Phipson replied:
Ah, got it! Sorry, I misunderstood the question. I'll take a pass on it and update this accordingly, thank you!

rr-codes commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,110 @@
+#if os(visionOS)
+
+internal import CoreRE
+import Combine
```
all of these should be internal import unless otherwise needed

rr-codes commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,110 @@
+/// A driver that maps all gesture updates to the specific transform we want for the specified StageMode behavior
+@MainActor
+@objc(WKStageModeInteractionDriver)
+public final class WKStageModeInteractionDriver: NSObject {
```
this is not the proper way to create a swift implementation. Please use `@objc @implementation`

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5568,6 +5569,29 @@ void WebPage::dragCancelled()
 #endif // ENABLE(DRAG_SUPPORT)
 
+#if ENABLE(MODEL_PROCESS)
+void WebPage::requestInteractiveModelElementAtPoint(WebCore::IntPoint clientPosition)
```
Nit - you can omit the `WebCore::` here, since there's a `using namespace` above.

whsieh commented on Source/WebKit/WebKitSwift/StageMode/WKStageModeBridging.h:
```diff
@@ -0,0 +1,50 @@
+- (void) interactionDidBegin:(simd_float4x4)transform NS_SWIFT_NAME(interactionDidBegin(_:));
+- (void) interactionDidUpdate:(simd_float4x4)transform NS_SWIFT_NAME(interactionDidUpdate(_:));
+- (void) interactionDidEnd NS_SWIFT_NAME(interactionDidEnd());
+- (void) operationDidUpdate:(WKStageModeOperation)operation NS_SWIFT_NAME(operationDidUpdate(_:));
```
```suggestion
- (instancetype)initWithInteractionTarget:(REEntityRef)target model:(REEntityRef)model container:(REEntityRef)container NS_SWIFT_NAME(init(with:model:container:));
- (instancetype)initWithInteractionTarget:(REEntityRef)target wkModel:(WKSRKEntity * _Nullable)wkModel container:(REEntityRef)container NS_SWIFT_NAME(init(with:wkModel:container:));
- (void)interactionDidBegin:(simd_float4x4)transform NS_SWIFT_NAME(interactionDidBegin(_:));
- (void)interactionDidUpdate:(simd_float4x4)transform NS_SWIFT_NAME(interactionDidUpdate(_:));
- (void)interactionDidEnd NS_SWIFT_NAME(interactionDidEnd());
- (void)operationDidUpdate:(WKStageModeOperation)operation NS_SWIFT_NAME(operationDidUpdate(_:));
```

whsieh commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,110 @@
+    
+    @objc(interactionDidEnd)
+    func interactionDidEnd() {
+        // TODO: rdar://141251753
```
nit - WebKit uses FIXME, rather than TODO. If possible, these should also reference bugzilla bugs instead of radars.

whsieh commented on Source/WebKit/WebKitSwift/RealityKit/RealityKitBridging.h:
```diff
@@ -70,3 +70,4 @@ typedef struct {
 NS_ASSUME_NONNULL_END
 
 #endif // defined(TARGET_OS_VISION) && TARGET_OS_VISION
+
```
Nit - stray whitespace

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -14181,6 +14317,26 @@ - (void)_dismissContactPickerWithContacts:(NSArray *)contacts
+{
+    return @{
+        @"awaitingResult" : _stageModeSession.state == WebKit::StageModeSessionState::Preparing ? @YES : @NO,
+        @"hitTestSuccessful" : _stageModeSession.elementID.has_value() ? @YES : @NO,
```
```suggestion
        @"awaitingResult" : @(_stageModeSession.state == WebKit::StageModeSessionState::Preparing),
        @"hitTestSuccessful" : @(_stageModeSession.elementID.has_value()),
```

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -14181,6 +14317,26 @@ - (void)_dismissContactPickerWithContacts:(NSArray *)contacts
+}
+
+#if ENABLE(MODEL_PROCESS)
+- (NSDictionary *) _stageModeInfoForTesting
```
```suggestion
- (NSDictionary *)_stageModeInfoForTesting
```

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    FloatPoint adjustedClientPositionAsFloatPoint(clientPosition);
+    frame->nodeRespondingToClickEvents(clientPosition, adjustedClientPositionAsFloatPoint);
+    IntPoint adjustedClientPosition = roundedIntPoint(adjustedClientPositionAsFloatPoint);
+    IntPoint adjustedGlobalPosition = frame->view()->windowToContents(adjustedClientPosition);
```
```suggestion
    auto adjustedClientPosition = roundedIntPoint(adjustedClientPositionAsFloatPoint);
    auto adjustedGlobalPosition = frame->view()->windowToContents(adjustedClientPosition);
```

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    auto documentPoint = frame->view() ? frame->view()->windowToContents(syntheticMousePressEvent.position()) : syntheticMousePressEvent.position();
+    auto hitTestedMouseEvent = document->prepareMouseEvent(hitType, documentPoint, syntheticMousePressEvent);
+
+    auto targetElement = hitTestedMouseEvent.hitTestResult().protectedTargetElement().get();
```
```suggestion
    RefPtr targetElement = hitTestedMouseEvent.hitTestResult().targetElement();
```

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+
+    auto targetElement = hitTestedMouseEvent.hitTestResult().protectedTargetElement().get();
+    if (targetElement && is<HTMLModelElement>(targetElement)) {
+        HTMLModelElement *modelElement = downcast<HTMLModelElement>(targetElement);
```
```suggestion
    if (RefPtr modelElement = dynamicDowncast<HTMLModelElement>(targetElement)) {
```

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    if (!element || !is<HTMLModelElement>(element))
+        return;
+
+    auto modelElement = downcast<HTMLModelElement>(element);
```
This can also use `dynamicDowncast`.

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+
+void EventHandler::stageModeSessionDidEnd(std::optional<WebCore::ElementIdentifier> elementID)
+{
+    if (!elementID.has_value())
```
```suggestion
    if (!elementID)
```

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    if (!elementID.has_value())
+        return;
+
+    RefPtr element = Element::fromIdentifier(elementID.value());
```
```suggestion
    RefPtr element = Element::fromIdentifier(*elementID);
```

pxlcoder commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    if (!element || !is<HTMLModelElement>(element))
+        return;
+
+    auto modelElement = downcast<HTMLModelElement>(element);
```
`dynamicDowncast`

pxlcoder commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -463,6 +464,25 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    // We want to first make sure that the model itself is properly positioned inside the portal
+    // Then, we a wrapper around the containerEntity for stage mode related interactions and transformations
+    // This ensures that the transforms applied on the model don't conflict with model-specific animations/JS transforms
+    REPtr<REEntityRef> interactionTarget = adoptRE(REEntityCreate());
```
```suggestion
    REPtr interactionTarget = adoptRE(REEntityCreate());
```

achan00 replied:
Also, we need to make sure this entity gets cleaned up like the other entities we created in the `ModelProcessModelPlayerProxy` destructor.

achan00 replied:
Can assign directly to `m_interactionContainerEntity` and remove the `interactionTarget` variable?

pxlcoder commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -702,6 +739,22 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+        break;
+    }
+
+    default: {
```
Prefer listing out all enum values over using `default`. This ensures that in the future, people think about how new values should behave when adding them.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
```diff
@@ -45,3 +45,5 @@ SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKTextExtractionTextItem)
 SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKTextExtractionImageItem)
 SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKSLinearMediaSpatialVideoMetadata)
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKStageModeInteractionDriver)
+
```
```suggestion
```

pxlcoder commented on Source/WebKit/UIProcess/Model/StageModeInteractionState.h:
```diff
@@ -0,0 +1,51 @@
+struct StageModeSession {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
```
Not need for `public` given it's a `struct`.

Phipson replied:
No I think I do need this, because otherwise I get a build error that I can't access the non-public members inside the struct. I see this approach being used in other structs in the codebase (e.g. ModelPresentationManagerProxy.h)

rr-codes replied:
shouldn't this be using WTF_MAKE_STRUCT_FAST_ALLOCATED and then the public can be omitted

pxlcoder commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -41,6 +41,7 @@
 #include <WebCore/DigitalCredentialsResponseData.h>
 #include <WebCore/DragActions.h>
 #include <WebCore/EditorClient.h>
+#include <WebCore/ElementTargetingTypes.h>
```
Can we forward declare rather than importing the header in this file?

whsieh replied:
I'm not sure we can forward declare `ElementIdentifier` here without redeclaring the whole thing, since it's just:

```
enum class ElementIdentifierType { };
using ElementIdentifier = ObjectIdentifier<ElementIdentifierType>;
```

That said, this shouldn't be pulling in `ElementIdentifier.h` through `ElementTargetingTypes.h`; instead, it should just include `ElementIdentifier.h` directly:

```suggestion
#include <WebCore/ElementIdentifier.h>
```

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -29,6 +29,7 @@
 #include "MessageReceiver.h"
 #include <WebCore/BoxExtents.h>
 #include <WebCore/CryptoKeyData.h>
+#include <WebCore/ElementTargetingTypes.h>
```
Prefer forward declaration if possible.

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1173,6 +1174,13 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 #endif // PLATFORM(IOS_FAMILY)
 
+#if ENABLE(MODEL_PROCESS)
+    void requestInteractiveModelElementAtPoint(const WebCore::IntPoint);
```
No need for `const`.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11236,6 +11316,62 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+    if (_stageModeSession.state != WebKit::StageModeSessionState::Preparing)
+        return;
+
+    if (elementID.has_value())
```
```suggestion
    if (elementID)
```

pxlcoder commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -20722,27 +20744,31 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = BCB86F4B116AAACD00CE20B7 /* WebKit.xcconfig */;
 			buildSettings = {
+				COMPILER_INDEX_STORE_ENABLE = NO;
```
Need to undo all of these.

pxlcoder commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -2242,6 +2242,102 @@ static BOOL isCompletelyWhite(UIImage *image)
 
+TEST(DragAndDropTests, IgnoreHitTestStageModeModel)
+{
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
Use explicit `RetainPtr` over `auto`. Multiple instances in this method.

pxlcoder commented on Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
```diff
@@ -462,6 +462,21 @@ - (void)simulateAllTouchesCanceled:(NSNotification *)notification
+    [_webView _simulatePanGestureBeginAtPoint:hitLocation];
+}
+
+- (bool)awaitingStageModeHitResult
```
```suggestion
- (BOOL)awaitingStageModeHitResult
```

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -481,6 +481,27 @@ void HTMLModelElement::didFinishEnvironmentMapLoading(bool succeeded)
+
+void HTMLModelElement::beginStageModeTransform(WebCore::TransformationMatrix transform)
+{
+    m_modelPlayer->beginStageModeTransform(transform);
```
Should check `m_modelPlayer` for `null`?

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -481,6 +481,27 @@ void HTMLModelElement::didFinishEnvironmentMapLoading(bool succeeded)
+
+void HTMLModelElement::updateStageModeTransform(WebCore::TransformationMatrix transform)
+{
+    m_modelPlayer->updateStageModeTransform(transform);
```
Ditto.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
```diff
@@ -168,7 +172,11 @@ class ModelProcessModelPlayerProxy final
+    REPtr<REEntityRef> m_interactionContainerEntity;
+    RetainPtr<WKStageModeInteractionDriver> m_stageModeInteractionDriver;
     WebCore::StageModeOperation m_stageModeOperation { WebCore::StageModeOperation::None };
+    void applyStageModeOperationToDriver();
```
Nit: group this method declaration with the other private methods.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -463,6 +464,25 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
     computeTransform();
     updateTransform();
+
+    // We want to first make sure that the model itself is properly positioned inside the portal
```
Can we move the logic to set up this new container entity before setting up the transform, opacity, and animation?

Phipson replied:
Deferred this to https://bugs.webkit.org/show_bug.cgi?id=287820 to simplify PR

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
```diff
@@ -532,6 +532,22 @@ - (void)_doAfterNextVisibleContentRectAndStablePresentationUpdate:(void (^)(void
+    [_contentView _simulatePanGestureUpdateAtPoint:hitPoint];
+}
+
+- (NSDictionary *) _stageModeInfoForTesting
```
```suggestion
- (NSDictionary *)_stageModeInfoForTesting
```

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
```diff
@@ -532,6 +532,22 @@ - (void)_doAfterNextVisibleContentRectAndStablePresentationUpdate:(void (^)(void
+- (void)_simulatePanGestureBeginAtPoint:(CGPoint)hitPoint
+{
+    [_contentView _simulatePanGestureBeginAtPoint:hitPoint];
+
```
```suggestion
```

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -862,6 +871,14 @@ FOR_EACH_PRIVATE_WKCONTENTVIEW_ACTION(DECLARE_WKCONTENTVIEW_ACTION_FOR_WEB_VIEW)
 
+-(void)panGestureDidBeginAtPoint:(CGPoint)inputPoint;
+-(void)panGestureDidUpdateWithPoint:(CGPoint)inputPoint;
+-(void)panGestureDidEnd;
```
Are these only called from within `WKContentViewInteraction.mm`? If so, they don't need to be declared in this header.

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -992,6 +1009,10 @@ FOR_EACH_PRIVATE_WKCONTENTVIEW_ACTION(DECLARE_WKCONTENTVIEW_ACTION_FOR_WEB_VIEW)
 
+- (void)_simulatePanGestureBeginAtPoint:(CGPoint)hitPoint;
+- (void)_simulatePanGestureUpdateAtPoint:(CGPoint)hitPoint;
+- (NSDictionary *) _stageModeInfoForTesting;
```
```suggestion
- (NSDictionary *)_stageModeInfoForTesting;
```

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -1372,6 +1376,14 @@ - (void)setUpInteraction
     [self addGestureRecognizer:_singleTapGestureRecognizer.get()];
 
+#if ENABLE(MODEL_PROCESS)
+    _singleTouchPanGestureRecognizer = adoptNS([[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(_singleTouchPanGestureRecognized:)]);
```
As I understand it, the `_singleTouchPanGestureRecognizer` is only used for this new model interaction — can we give it a more specific name (like `_modelInteractionPanGestureRecognizer`)?

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -3852,6 +3902,36 @@ - (void)_didCompleteSyntheticClick
+    ASSERT(gestureRecognizer == _singleTouchPanGestureRecognizer);
+
+    if (![self isFirstResponder]) {
+        [self startDeferringInputViewUpdates:WebKit::InputViewUpdateDeferralSource::TapGesture];
```
Hm.. is this right? It's unclear:
- Why we need to defer keyboard bringup for this gesture
- When this flag gets _un_-set, for this gesture
- Why it keys off of `TapGesture`, when this is a pan gesture

Phipson replied:
Oh, I misinterpreted the above block. I was of the impression that this was necessary to cancel text related input and reset the responder chain. I don't think that's relevant here since WebKit already has text gesture-related handling. Will remove

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11236,6 +11316,57 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+
+-(void)updateStageModeSessionTransformWithPoint:(CGPoint)inputPoint {
+    WebCore::TransformationMatrix tf;
+    tf.makeIdentity();
```
I think the default ctor already creates the identity matrix.

Also, `tf` should be `transform` or `matrix` (we prefer not to use letter abbreviations for variable names in WebKit code)

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -14181,6 +14312,32 @@ - (void)_dismissContactPickerWithContacts:(NSArray *)contacts
+- (NSDictionary *)_stageModeInfoForTesting
+{
+#if ENABLE(MODEL_PROCESS)
+
```
```suggestion
```

achan00 commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +699,82 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    FloatPoint adjustedClientPositionAsFloatPoint(clientPosition);
+    frame->nodeRespondingToClickEvents(clientPosition, adjustedClientPositionAsFloatPoint);
+    auto adjustedClientPosition = roundedIntPoint(adjustedClientPositionAsFloatPoint);
+    auto adjustedGlobalPosition = frame->view()->windowToContents(adjustedClientPosition);
```
Can store `frame->view()` into a `RefPtr` and check for null earlier in the method.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -463,6 +464,25 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+    // We want to first make sure that the model itself is properly positioned inside the portal
+    // Then, we a wrapper around the containerEntity for stage mode related interactions and transformations
+    // This ensures that the transforms applied on the model don't conflict with model-specific animations/JS transforms
+    REPtr interactionTarget = adoptRE(REEntityCreate());
```
Can we assign to `m_interactionContainerEntity` here already?

Following what Aleksei has done with `m_containerEntity` and `m_hostingEntity`, we should make sure we call `REEntityRemoveFromSceneOrParent` on `m_interactionContainerEntity ` in the destructor.

Phipson replied:
Deferred this to https://bugs.webkit.org/show_bug.cgi?id=287820 to simplify PR

achan00 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -349,6 +352,7 @@ struct ImageAnalysisContextMenuActionData {
     RetainPtr<WKHighlightLongPressGestureRecognizer> _highlightLongPressGestureRecognizer;
     RetainPtr<UILongPressGestureRecognizer> _longPressGestureRecognizer;
     RetainPtr<WKSyntheticTapGestureRecognizer> _doubleTapGestureRecognizer;
+    RetainPtr<UIPanGestureRecognizer> _singleTouchPanGestureRecognizer;
```
Should this be guarded with `ENABLE(MODEL_PROCESS)`?

achan00 commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -0,0 +1,110 @@
+    // MARK: ObjC Exposed API
+    @objc(initWithInteractionTarget:model:container:)
+    convenience init(with entity: REEntityRef, model: REEntityRef, container: REEntityRef) {
+        let interactionContainer = Entity.__fromCore(__EntityRef.__fromCore(entity))
```
Do we need this version that takes in core `REEntityRef`? Maybe simpler to just have one that takes in all `WKSRKEntity`. The caller can create `WKSRKEntity` from `REEntityRef` with `-[WKSRKEntity initWithCoreEntity:]`.

Phipson replied:
Deferred this to https://bugs.webkit.org/show_bug.cgi?id=287820 to simplify PR

achan00 commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -2242,6 +2242,104 @@ static BOOL isCompletelyWhite(UIImage *image)
+    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    for (_WKFeature *feature in [WKPreferences _features]) {
+        if ([feature.key isEqualToString:@"ModelElementEnabled"])
+            [[configuration preferences] _setEnabled:YES forFeature:feature];
```
We should also make sure `ModelProcessEnabled` feature is enabled.

achan00 commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -2242,6 +2242,104 @@ static BOOL isCompletelyWhite(UIImage *image)
+        [task didReceiveResponse:response];
+        [task didReceiveData:modelData.get()];
+        [task didFinish];
+    }];
```
I updated the `CanStartDragOnModel` to load from a file. You can try that also here if you like.

achan00 commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -2242,6 +2242,104 @@ static BOOL isCompletelyWhite(UIImage *image)
+    RetainPtr configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    for (_WKFeature *feature in [WKPreferences _features]) {
+        if ([feature.key isEqualToString:@"ModelElementEnabled"])
+            [[configuration preferences] _setEnabled:YES forFeature:feature];
```
Ditto about enabling `ModelProcessEnabled` feature.

achan00 commented on Source/WebKit/UIProcess/Model/StageModeInteractionState.h:
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2007-2025 Apple Inc. All rights reserved.
```
Should just be `2025`

achan00 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11264,6 +11345,50 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
 
+#pragma mark - Model Interaction Support
+#if ENABLE(MODEL_PROCESS)
+-(void)modelInteractionPanGestureDidBeginAtPoint:(CGPoint)inputPoint {
```
There should be a space between `-` and `(void)`. Ditto for the other methods in this file.

achan00 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -863,6 +874,13 @@ FOR_EACH_PRIVATE_WKCONTENTVIEW_ACTION(DECLARE_WKCONTENTVIEW_ACTION_FOR_WEB_VIEW)
 #endif
 
+#if ENABLE(MODEL_PROCESS)
+-(void)modelInteractionPanGestureDidBeginAtPoint:(CGPoint)inputPoint;
```
Nit: need space between `-` and `(void)`

achan00 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -602,6 +608,11 @@ struct ImageAnalysisContextMenuActionData {
 #endif
+
+#if ENABLE(MODEL_PROCESS)
+    WebKit::StageModeSession _stageModeSession;
```
Another option is to make this a `std::optional` - so it only has a value if a stage mode session has started (and maybe we wouldn't need the `StageModeSessionState::Idle`).

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -53,6 +53,7 @@
 #import "WAKView.h"
 #import "WAKWindow.h"
 #import "WebEvent.h"
+#import <WebCore/ElementTargetingTypes.h>
```
This is the wrong header to import; should be `ElementIdentifier.h`

Also, we should never need to import `<WebCore/….h>` from within WebCore source files.

whsieh commented on Source/WebCore/Modules/model-element/ModelPlayer.cpp:
```diff
@@ -118,6 +118,19 @@ void ModelPlayer::setHasPortal(bool)
 {
 }
+
+void ModelPlayer::beginStageModeTransform(WebCore::TransformationMatrix)
```
```suggestion
void ModelPlayer::beginStageModeTransform(const TransformationMatrix&)
```

No need for `WebCore::` namespacing, when you're already in a WebCore namespace.

whsieh commented on Source/WebCore/Modules/model-element/ModelPlayer.h:
```diff
@@ -98,6 +98,9 @@ class WEBCORE_EXPORT ModelPlayer : public RefCounted<ModelPlayer> {
     virtual void setHasPortal(bool);
     virtual void setStageMode(WebCore::StageModeOperation);
+    virtual void beginStageModeTransform(WebCore::TransformationMatrix);
+    virtual void updateStageModeTransform(WebCore::TransformationMatrix);
```
```suggestion
    virtual void beginStageModeTransform(const TransformationMatrix&);
    virtual void updateStageModeTransform(const TransformationMatrix&);
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +700,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    RefPtr targetElement = hitTestedMouseEvent.hitTestResult().targetElement();
+    if (RefPtr modelElement = dynamicDowncast<HTMLModelElement>(targetElement)) {
+        if (modelElement->supportsStageModeInteraction()) {
+            WebCore::TransformationMatrix tf = TransformationMatrix::identity;
```
```suggestion
            auto transform = TransformationMatrix::identity;
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +700,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    if (!element || !is<HTMLModelElement>(element))
+        return;
+
+    RefPtr modelElement = dynamicDowncast<HTMLModelElement>(element);
```
```suggestion
    if (!element)
        return;

    RefPtr modelElement = dynamicDowncast<HTMLModelElement>(element);
    if (!modelElement)
        return;
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +700,83 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+        return;
+
+    RefPtr element = Element::fromIdentifier(*elementID);
+    if (!element || !is<HTMLModelElement>(element))
```
Ditto.

whsieh commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
```diff
@@ -131,6 +132,9 @@ class ModelProcessModelPlayerProxy final
     void setHasPortal(bool) final;
     void setStageMode(WebCore::StageModeOperation) final;
+    void beginStageModeTransform(WebCore::TransformationMatrix) final;
+    void updateStageModeTransform(WebCore::TransformationMatrix) final;
```
```suggestion
    void beginStageModeTransform(const TransformationMatrix&) final;
    void updateStageModeTransform(const TransformationMatrix&) final;
```

whsieh commented on Source/WebKit/UIProcess/Model/StageModeInteractionState.h:
```diff
@@ -0,0 +1,45 @@
+
+#if ENABLE(MODEL_PROCESS)
+
+#import <WebCore/ElementTargetingTypes.h>
```
This is the wrong header.

whsieh commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -67,6 +67,7 @@
 #import <WebCore/Cursor.h>
 #import <WebCore/DOMPasteAccess.h>
 #import <WebCore/DictionaryLookup.h>
+#import <WebCore/ElementTargetingTypes.h>
```
Ditto.

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -120,6 +120,7 @@
 #import <WebCore/CompositionHighlight.h>
 #import <WebCore/DOMPasteAccess.h>
 #import <WebCore/DataDetection.h>
+#import <WebCore/ElementTargetingTypes.h>
```
Ditto.

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11264,6 +11345,49 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+#if ENABLE(MODEL_PROCESS)
+- (void)modelInteractionPanGestureDidBeginAtPoint:(CGPoint)inputPoint {
+    if (!_stageModeSession) {
+        _stageModeSession = { };
```
This doesn't seem necessary — `_stageModeSession` is already nullopt

whsieh commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -69,6 +69,7 @@
 #import "WebProcessProxy.h"
 #import "WebScreenOrientationManagerProxy.h"
 #import <WebCore/AGXCompilerService.h>
+#import <WebCore/ElementTargetingTypes.h>
```
Ditto (wrong header)

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11264,6 +11345,49 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+
+// FIXME: https://bugs.webkit.org/show_bug.cgi?id=287601
+- (void)modelInteractionPanGestureDidUpdateWithPoint:(CGPoint)inputPoint {
+    if (_stageModeSession && !_stageModeSession.isPreparingForInteraction) {
```
It's actually a bit unclear how this compiles, since `_stageModeSession` is a `std::optional<>`…

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11264,6 +11345,49 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+    _stageModeSession.elementID = elementID;
+}
+
+- (void)cleanUpStageModeSessionState {
```
```suggestion
- (void)cleanUpStageModeSessionState
{
```

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11264,6 +11345,49 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+    [self cleanUpStageModeSessionState];
+}
+
+- (void)didReceiveInteractiveModelElement:(std::optional<WebCore::ElementIdentifier>)elementID {
```
```suggestion
- (void)didReceiveInteractiveModelElement:(std::optional<WebCore::ElementIdentifier>)elementID
{
```

(And in other methods in this file)

achan00 commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11264,6 +11345,54 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+    _stageModeSession = std::nullopt;
+}
+
+- (void)updateStageModeSessionTransformWithPoint:(CGPoint)inputPoint
```
Does this need to be a separate method? From code searching, I've only seen this called in `-modelInteractionPanGestureDidUpdateWithPoint:`.

achan00 commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -186,6 +186,7 @@ class HTMLPlugInElement;
 class HTMLSelectElement;
 class HTMLTextFormControlElement;
 class HTMLVideoElement;
+class HTMLModelElement;
```
Is this needed in the header? I didn't notice references to `HTMLModelElement` in the new methods being added to this header file.

achan00 commented on Source/WebKit/UIProcess/Model/StageModeInteractionState.h:
```diff
@@ -0,0 +1,45 @@
+
+public:
+    bool isPreparingForInteraction { true };
+    simd_float4x4 transform;
```
Should we initialize this to `matrix_identity_float4x4`?

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5576,6 +5577,29 @@ void WebPage::dragCancelled()
+        localMainFrame->eventHandler().stageModeSessionDidUpdate(elementID, transform);
+}
+
+void WebPage::stageModeSessionDidEnd(std::optional<WebCore::ElementIdentifier> elementID)
```
```suggestion
void WebPage::stageModeSessionDidEnd(std::optional<ElementIdentifier> elementID)
```

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11264,6 +11345,49 @@ - (void)dropInteraction:(UIDropInteraction *)interaction sessionDidEnd:(id<UIDro
+    if (!_stageModeSession || !_stageModeSession->isPreparingForInteraction)
+        return;
+
+    _stageModeSession->isPreparingForInteraction = !elementID.has_value();
```
```suggestion
    _stageModeSession->isPreparingForInteraction = !elementID;
```

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1177,6 +1181,13 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
+#if ENABLE(MODEL_PROCESS)
+    void requestInteractiveModelElementAtPoint(WebCore::IntPoint);
+    void didReceiveInteractiveModelElement(std::optional<WebCore::ElementIdentifier>);
+    void stageModeSessionDidUpdate(std::optional<WebCore::ElementIdentifier>, WebCore::TransformationMatrix);
```
```suggestion
    void stageModeSessionDidUpdate(std::optional<WebCore::ElementIdentifier>, const WebCore::TransformationMatrix&);
```

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3853,6 +3853,23 @@ void WebPageProxy::setDragCaretRect(const IntRect& dragCaretRect)
+void WebPageProxy::stageModeSessionDidEnd(std::optional<WebCore::ElementIdentifier> elementID)
+{
+    send(Messages::WebPage::StageModeSessionDidEnd(elementID));
+}
```
```suggestion
void WebPageProxy::requestInteractiveModelElementAtPoint(const IntPoint clientPosition)
{
    send(Messages::WebPage::RequestInteractiveModelElementAtPoint(clientPosition));
}

void WebPageProxy::stageModeSessionDidUpdate(std::optional<ElementIdentifier> elementID, const TransformationMatrix& transform)
{
    send(Messages::WebPage::StageModeSessionDidUpdate(elementID, transform));
}

void WebPageProxy::stageModeSessionDidEnd(std::optional<ElementIdentifier> elementID)
{
    send(Messages::WebPage::StageModeSessionDidEnd(elementID));
}
```

whsieh commented on Source/WebKit/UIProcess/Model/StageModeInteractionState.h:
```diff
@@ -0,0 +1,49 @@
+public:
+    bool isPreparingForInteraction { true };
+    simd_float4x4 transform { matrix_identity_float4x4 };
+    std::optional<WebCore::ElementIdentifier> elementID { std::nullopt };
```
Hm... I _think_ this shouldn't be necessary, since `std::optional` will initialize to `nullopt` by default:

```suggestion
    std::optional<WebCore::ElementIdentifier> elementID;
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +700,87 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+
+    modelElement->endStageModeInteraction();
+}
+
```
```suggestion
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +700,87 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    return std::nullopt;
+}
+
+void EventHandler::stageModeSessionDidUpdate(std::optional<WebCore::ElementIdentifier> elementID, WebCore::TransformationMatrix transform)
```
```suggestion
void EventHandler::stageModeSessionDidUpdate(std::optional<ElementIdentifier> elementID, const TransformationMatrix& transform)
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +700,87 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
+    modelElement->updateStageModeTransform(transform);
+}
+
+void EventHandler::stageModeSessionDidEnd(std::optional<WebCore::ElementIdentifier> elementID)
```
```suggestion
void EventHandler::stageModeSessionDidEnd(std::optional<ElementIdentifier> elementID)
```

whsieh commented on Source/WebCore/page/ios/EventHandlerIOS.mm:
```diff
@@ -693,6 +700,87 @@ static IntPoint adjustAutoscrollDestinationForInsetEdges(IntPoint autoscrollPoin
 }
 
+#if ENABLE(MODEL_PROCESS)
+std::optional<WebCore::ElementIdentifier> EventHandler::requestInteractiveModelElementAtPoint(const IntPoint& clientPosition)
```
```suggestion
std::optional<ElementIdentifier> EventHandler::requestInteractiveModelElementAtPoint(const IntPoint& clientPosition)
```

whsieh commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -153,6 +153,10 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
     void setEnvironmentMap(const URL&);
+    WEBCORE_EXPORT bool supportsStageModeInteraction() const;
+    WEBCORE_EXPORT void beginStageModeTransform(WebCore::TransformationMatrix);
+    WEBCORE_EXPORT void updateStageModeTransform(WebCore::TransformationMatrix);
```
```suggestion
    WEBCORE_EXPORT void beginStageModeTransform(TransformationMatrix);
    WEBCORE_EXPORT void updateStageModeTransform(TransformationMatrix);
```

whsieh commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -481,6 +481,29 @@ void HTMLModelElement::didFinishEnvironmentMapLoading(bool succeeded)
+        m_modelPlayer->beginStageModeTransform(transform);
+}
+
+void HTMLModelElement::updateStageModeTransform(WebCore::TransformationMatrix transform)
```
```suggestion
    return stageMode() != StageModeOperation::None;
}

void HTMLModelElement::beginStageModeTransform(const TransformationMatrix& transform)
{
    if (m_modelPlayer)
        m_modelPlayer->beginStageModeTransform(transform);
}

void HTMLModelElement::updateStageModeTransform(const TransformationMatrix& transform)
```

whsieh commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -350,6 +354,12 @@ class EventHandler final : public CanMakeCheckedPtr<EventHandler> {
+#if ENABLE(MODEL_PROCESS)
+    WEBCORE_EXPORT std::optional<WebCore::ElementIdentifier> requestInteractiveModelElementAtPoint(const IntPoint& clientPosition);
+    WEBCORE_EXPORT void stageModeSessionDidUpdate(std::optional<WebCore::ElementIdentifier>, WebCore::TransformationMatrix);
+    WEBCORE_EXPORT void stageModeSessionDidEnd(std::optional<WebCore::ElementIdentifier>);
```
```suggestion
    WEBCORE_EXPORT std::optional<ElementIdentifier> requestInteractiveModelElementAtPoint(const IntPoint& clientPosition);
    WEBCORE_EXPORT void stageModeSessionDidUpdate(std::optional<ElementIdentifier>, const TransformationMatrix&);
    WEBCORE_EXPORT void stageModeSessionDidEnd(std::optional<ElementIdentifier>);
```

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -164,7 +168,6 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
 #if ENABLE(ARKIT_INLINE_PREVIEW_MAC)
     WEBCORE_EXPORT String inlinePreviewUUIDForTesting() const;
 #endif
-
```
How come this newline got removed?

Phipson committed:
Add Drag Gesture for StageMode for Model View
https://bugs.webkit.org/show_bug.cgi?id=287048
rdar://137779076

Reviewed by Wenson Hsieh and Ada Chan.

This PR is the second part of 3 code changes used to introduce stagemode interactions to the model view.
Specifically, this code change adds a UIPanGestureRecognizer that will be used to update the entityTransform
for the model view in rdar://141251753. When the stagemode attribute is set to orbit, and we have hit an
HTMLModelElement, the gesture will feed transform updates to the model player, which will use those transforms
to update the transform of the model. We wrap the model inside an interaction container, to separate gesture-
based transform updates from JavaScript/Animation-driven transform updates. Also, transform supplied by the
gesture will be fed into a StageModeInteractionDriver that will alter the transform of the model, depending
on the stagemode attribute that has been set for the element.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::supportsStageModeInteraction const):
(WebCore::HTMLModelElement::updateStageModeTransform):
(WebCore::HTMLModelElement::endStageModeInteraction):
- Add API calls to determine if stagemode is defined for the model; to update the model via the model player;
and end the interaction via the model player once the gesture terminates

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::HTMLModelElement::supportsStageModeInteraction const):
(WebCore::HTMLModelElement::beginStageModeTransform):
(WebCore::HTMLModelElement::updateStageModeTransform):
(WebCore::HTMLModelElement::endStageModeInteraction):
- Add API calls to determine if stagemode is defined for the model; to update the model via the model player;
and end the interaction via the model player once the gesture terminates

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::beginStageModeTransform):
(WebCore::ModelPlayer::updateStageModeTransform):
(WebCore::ModelPlayer::endStageModeInteraction):
- Defines the virtual and default behavior of the model player to handle stage mode updates from the gesture;
to be implemented by the ModelProcessModelPlayer and ModelProcessModelPlayerProxy

* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::requestInteractiveModelElementAtPoint):
(WebCore::EventHandler::stageModeSessionDidUpdate):
(WebCore::EventHandler::stageModeSessionDidEnd):
- Handles hit-testing and querying the HTMLModelElement (if available) to perform the stagemode interaction

* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.messages.in:
* Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
(WebKit::ModelProcessModelPlayerProxy::didFinishLoading):
(WebKit::ModelProcessModelPlayerProxy::beginStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::updateStageModeTransform):
(WebKit::ModelProcessModelPlayerProxy::endStageModeInteraction):
- Implements the transform updates directly on the model in the model process, to be done in https://bugs.webkit.org/show_bug.cgi?id=287820

* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:
* Source/WebKit/UIProcess/Model/StageModeInteractionState.h: Added.
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestInteractiveModelElementAtPoint):
(WebKit::WebPageProxy::stageModeSessionDidUpdate):
(WebKit::WebPageProxy::stageModeSessionDidEnd):
- Passes calls from the WKContentViewInteraction gesture to the WebPage in the Web Process

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didReceiveInteractiveModelElement):
- Passes the result of the initial query when the gesture begins back to the content view to update its interaction state

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView cleanUpInteraction]):
(-[WKContentView _removeDefaultGestureRecognizers]):
(-[WKContentView _addDefaultGestureRecognizers]):
(-[WKContentView gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:]):
(-[WKContentView gestureRecognizerShouldBegin:]):
(-[WKContentView _modelInteractionPanGestureRecognized:]):
(-[WKContentView modelInteractionPanGestureDidBeginAtPoint:]):
(-[WKContentView modelInteractionPanGestureDidUpdateWithPoint:]):
(-[WKContentView modelInteractionPanGestureDidEnd]):
(-[WKContentView didReceiveInteractiveModelElement:]):
(-[WKContentView cleanUpStageModeSessionState]):
(-[WKContentView updateStageModeSessionTransformWithPoint:]):
- Defines functions to add the model-specific UIPanGestureRecognizer and maintain an internal interaction state of the stagemode
interaction

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::didReceiveInteractiveModelElement):
- Receives the hit-testing result from the Web Process and routes calls to the PageClient

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::beginStageModeTransform):
(WebKit::ModelProcessModelPlayer::updateStageModeTransform):
(WebKit::ModelProcessModelPlayer::endStageModeInteraction):
- Routes calls from the HTMLModelElement to the ModelProcessModelPlayerProxy

* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestInteractiveModelElementAtPoint):
(WebKit::WebPage::stageModeSessionDidUpdate):
(WebKit::WebPage::stageModeSessionDidEnd):
- Receives the stagemode message from the UI Process and hands it off to the EventHandler

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::positionInformation):
- Adds state information about whether there is a model element with stagemode enabled at the position we are hit-testing

Canonical link: https://commits.webkit.org/290687@main
